docs/nsm_api.md

```diff
@@ -0,0 +1,263 @@
+
+**nse\_provider\_namespace** specifies NSE&#39;s kubernetes object namespace, NSM used it to differentiate between multiple NSEs providing the same service
+
+**socket\_location** informs NSM about linux named socket it has to use to communicate with NSE for connection requests
+
+### Local NSM to remote NSM not yet implemented
+
+When NSM local to NSM client discovers that NSE providing requested Network Service is not local, **network_service_host** in NSE custom resource object does not match the local NSM name, local NSM attempts to proxy client's request to remote NSM. gRPC over well known TCP socket is used for NSM to NSM communication. This method supports as "in-cluster" mode when NSM pod's DNS named is used as "out-of-cluster" when routable IP of external NSM is used to establish TCP connection. In order to facilitate NSM discovery for "in-cluster" mode, each NSM creates a kubernetes **Service** object with matching to NSM daemonset name. For each **Service**, a corresponding DNS entry is automatically created
+making discovery of any "in-cluster" NSM flexible and independent of NSM ip address changes.   
+
+- Proxy Connection request
+
+In this message Client&#39;s local NSM proxying Network Service request to NSE&#39;s local NSM.  2 NSMs need to agree upon a tunneling technology supported by both, hence a list of tunnel types is a part of the request message.
+
+```proto
+message ProxyConnectionRequest {
+   string request_id = 1;
+   string network_service_name = 2;
+   string nse_provider_name = 3;
+   repeated common.Tunnels tunnel_type = 4;
+}
+```
+**Where:**
+
+**request\_id** is NSM POD UID which is unique and immutable identifier existing throughout POD&#39;s life.
+
+**network\_service\_name** represents a network service/application, a client pod desires to connect to.
+
+**nse\_provider\_name** specifies NSE name, NSM used it to differentiate between multiple NSEs providing the same service
+
+**tunnel\_type list** defines supported by NSM POD tunnel types.
+
+- Proxy Connection reply
+
+After completing control plane signalling, programming of the dataplane for NSE pod and setting up a NSM to NSM tunnel endpoint, NSE&#39;s NSM responds to Client&#39;s NSM with Proxy Connection reply message.
+
+```proto
+message ProxyConnectionReply {
+   bool accepted = 1;
+   string admission_error = 2;
+   common.Tunnel tunnel = 4;
+}
+```
+**Where:**
+
+**accepted** true will indicate that the connection is accepted, otherwise false
+
+**admission\_error** will provide details why connection was refused.
+
+**interface** will indicate the selected/negotiated tunnel type
+
+### Local NSM to NSE
+
+NSM communicates with NSE to request a services on behalf of some NSM client. The communication occurs over a linux named socket exposed and advertised by NSE in its Endpoint Advertise Request. NSE replies NSM with some parameters NSM needs to program dataplane connection.
```

Comment:
If we talk about the remote NSM/NSE use case, i'm not sure the connection between NSM/NSE will occur over a Linux socket.

Reply:
in Remote NSM scenario I suspect it will be up to a external NSM/NSE developer to decide, all we need is Network Service Endpoint CR (custom resource) created with correct/reachable TCP/IP address so "in-cluster" NSM could build tcp/ip session.

BTW thanks so much for review and questions!!